### DISCUSSION TITLE
Field component

### DISCUSSION THREAD

**Discussion by asherccohen** [tayv/Project-Bubblegum#48] [Ideas]
The Field component is an abstraction over Controller that can help with enhancing its logic and simplify conditional rendering.

Here's a simple example of the API:

```
<Field
name="firstName"
defaultValue="tay"
validationRules={{required: true}}
...other basic field props
...custom component's props
>
   <CustomComponent />
</Field>

```

With this mechanism all of the controller's props can be merged with the custom component's.

With a bit of work this can also include conditional logic:

```
<WatchField
name="firstName"
defaultValue="tay"
validationRules={{required: true}}
...other basic field props
...custom component's props
conditional={{
name:"otherField" 
//name of the field to track
value: "valueOfOtherField" 
// render when this value matches
}}
>
   <CustomComponent />
</Field>

```

What's important to note is that "watch" will be local to the field, and not shared state that will cause the full form to re-render.
Using "watch" in an object like in your example causes all fields to be tracked, therefore all of the below components to re-render whenever one field changes.
Basically you're reinventing RHF formState, we need to avoid this.

**Comment by tayv** [comment]
Okay I here's my first pass at it. Seems to work in my test form. I treated these similar to my original Wrapper__ components.

**This is still rough but is it in line with what you're thinking?**

## Standard Field Component

**Overview**

1. Created a `Field.tsx` component that accepts basic field + custom component props.
2. Use React.cloneElement() in the Controller's render prop to render the child component
3. Called Field in the test form. It renders as expected.

<details> 
  <summary>Example: Field.tsx component</summary>

````
import Checkbox from '@designSystem/atoms/Checkbox'
import React, { FC, ReactElement } from 'react'
import { Controller, Control } from 'react-hook-form'

type FieldProps = {
  name: string
  defaultValue: any
  validationRules?: any
  control: Control
  children: ReactElement
}

const Field: FC<FieldProps> = ({
  name,
  defaultValue,
  validationRules,
  control,
  children,
  ...props
}) => {
  return (
    <Controller
      name={name}
      defaultValue={defaultValue}
      control={control}
      rules={validationRules}
      render={ ({ field: {...props } }) => (
        React.cloneElement(children, {
          ...props,
          ...children.props
        } )    
      ) }   
   />
  )
}

export default Field
````
</details>

<details>
  <summary>Example: Using Field in the test form</summary>

 ````
 <Field
    name="field-test-checkbox"
    defaultValue={true}
    control={control}
    >
    <Checkbox label="test-checkbox" />  // NOTE: Refactoring still needed so don't have to pass props at child level 
 </Field>

````
</details>


## Conditional WatchField Component

**Overview**
1. Same as `Field` component except added props for `watch` and `conditional`
2. Do a conditional check based on prop values in `WatchField` component. If it fails return `null`
3. Use `WatchField` in the test form. It appears to work as expected.

<details>
  <summary>Example: WatchField.tsx component</summary>
 
````
import Checkbox from '@designSystem/atoms/Checkbox'
import React, { FC, ReactElement } from 'react'
import { Controller, Control, UseFormWatch } from 'react-hook-form'

type FieldProps = {
  name: string
  defaultValue: any
  validationRules?: any
  control: Control
  children: ReactElement
  watch: any
  conditional: any
}

const WatchField: FC<FieldProps> = ({
  name,
  defaultValue,
  validationRules,
  control,
  children,
  watch,
  conditional,
  ...props
}) => {

  // Watch field and then check to if the field should be rendered
  const otherFieldValue = watch(conditional.name)

  if (otherFieldValue !== conditional.value) {
    return null
  }

  return (
    <Controller
      name={name}
      defaultValue={defaultValue}
      control={control}
      rules={validationRules}
      render={({ field: {...props } }) => (
        React.cloneElement(children, {
          ...props,
          ...children.props
        })
      ) } 
   />
  )
}

export default WatchField

 ````
</details>


<details>
 <summary>Example: Using WatchField in the test form</summary>
 
````
 <WatchField
   name="watchfield-test-checkbox"
   defaultValue={true}
   control={control}
   watch={watch}
   conditional={ { name: 'field-test-checkbox', value: false } }
   >
   <Checkbox label="watchfield-test-checkbox" />
 </WatchField>
 
````
</details>

**Reply by tayv** [reply]
Now on main branch: https://github.com/tayv/Project-Bubblegum/blob/main/pages/testForms/test-viscond-form.tsx#L168-L201

**Comment by asherccohen** [comment]
Yes, this is precisely what I imagined.
A few things to consider.

https://github.com/tayv/Project-Bubblegum/blob/97d72284cab5d9ffeafca2f75b04a4dde41d787d/designSystem/forms/FieldTest.tsx#L27
Here I can show you how to use some internal radix ui utils to replace cloneElement with asChild.

At some point you will want to give a label to your Field (which could also need to read the field's state)

```
const Field: FC<FieldProps> = ({
  name,
  defaultValue,
  validationRules,
  control,
  children,
  ...props
}) => {
  return (
    <Controller
      name={name}
      defaultValue={defaultValue}
      control={control}
      rules={validationRules}
      render={ ({ field: {...props } }) => (
      //LABEL
        React.cloneElement(children, {
          ...props,
          ...children.props
        })    
        //CAPTION
      ) }
      
   />

  )
}
```

To create this markup:
```

    <Form.Field>
      <Form.Label />
      <Form.Control />
      <Form.Message />
    </Form.Field>
```

The problem is that you can only clone 1 child and pass the props to it, so to achieve this you have 2 options:
- place the Label element directly in Controller
- share the methods and state from controller in a Context, create Label, Control, Message components that can consume the context

While we speak, radix has released a Form component, worth exploring if it matches your criteria, maybe you can replace RHF altogether (they have a lot less features though)

They basically offer the same DX out of the box:

```
<Form.Root>
    <Form.Field>
      <Form.Label />
      <Form.Control />
      <Form.Message />
      <Form.ValidityState />
    </Form.Field>

    <Form.Message />
    <Form.ValidityState />

    <Form.Submit />
  </Form.Root>
```

**Comment by tayv** [comment]
Just to make sure I understand correctly:

- cloneElement is brittle as it can only handle one child component at a time. For this reason using `asChild` 
- Your suggestion looks to me like the [Compound Component pattern](https://www.smashingmagazine.com/2021/08/compound-components-react/). Is this correct?
  - Are you suggesting I create my own Field component using that pattern?
> Here I can show you how to use some internal radix ui utils to replace cloneElement with asChild.
  - I don't fully understand how this would work

Re: Radix Form
- This looks really interesting and something I'd be interested in adopting. I'll have to spend some time digging into the documentation. Right now my biggest question mark would be how they handle visibility conditions since I'm now use rhf's `watch`

**Reply by asherccohen** [reply]
To clarify:

- using cloneElement directly has a lot of gotchas, radix has simplified all of that with a Slot component (it's internal to the lib, so we have to go and dig for it).
This helps with polymorphism (a component that can act as another one), it merged props correctly, has a few solid rules to avoid shooting your foot and keeps typescript happy.

- yes, this pattern is called Compound Component. Just like all other radix components, using composition is a lot more powerful than creating Config Components (I call the typical component with many props like that).

- the radix form is limited. I would stick with RHF (which also has released a Form component btw), but worth getting some ideas from them.

**Reply by tayv** [reply]
I think Slot is now exposed as a utility primitive: https://www.radix-ui.com/docs/primitives/utilities/slot

**Comment by tayv** [comment]
Still a work in progress but refactoring Field so that it uses context. Hopefully this aligns with your suggestions. Example:

```
<form onSubmit={handleSubmit(onSubmit)}>
  <Field
    name="firstName"
    defaultValue="jim"
    validationRules={{ required: 'First name is required' }}
    control={control}
  >
    <Label>First Name:</Label>
    <Control /> // Atomic form components like Select, Input, etc.
    <Message validationMessage="Please enter your first name" />
  </Field>
  <button type="submit">Submit</button>
</form>
```

Then update the child form components to receive context. Something like: ` const fieldContext = useContext(FieldContext)`

**Reply by asherccohen** [reply]
Cool.
A few things you'll notice while building it.

<Control>
This could anything from Select, Input, etc to fully custom components given as children to the Field.
It will take all the props from Controller (particularly "control", "value" and "onChange")

<Label> 
Depending on your styling approach, you might want the label to use some of the Field's state (Eg isValid to show a positive state, isError to show a negative state)

This means that when used with Controller (which gives you a render prop) you have two choices:
- keep the label encapsulated inside the Controller render
- add another Context to propagate the field state

<Message> or <Caption>
Same as Label

What I mean here is that the downside/tradeoff of using compound components is that at some point you'll have to nest Contexts.

I won't go into the details of using FormProvider to also propagate useForm's methods. That's the last thing I would worry about and it really depends how you want this API to work.
We'll expand after the first few iterations.

**Reply by tayv** [reply]
Here's my latest work. Types aren't done yet and still need to add in states but seems to work as expected so far:
https://github.com/tayv/Project-Bubblegum/blob/049ed592883a7a334246b946ef0518e6b6968d67/designSystem/forms/Field.tsx#L7-L55

I then would use it like this in a form:

https://github.com/tayv/Project-Bubblegum/blob/049ed592883a7a334246b946ef0518e6b6968d67/pages/docs/field-article.tsx#L79-L96

**Comment by tayv** [comment]
Here's my latest:

#### To do:
- [x] Refactor `FieldTest` to be a generic wrapper component that will make any atomic/molecule form input controlled 
- [x] Use the context API and compound component pattern in `Field` 
- [x] Add label for the input or group of inputs (Field.GroupLabel). This is different than individual labels for grouped option inputs like Checkboxes and Radio buttons.
- [x] Add Field.Tip message for clarifying text under the group label
- [x] Field.Control for passing the child form component and making it controlled via RHF's Controller
- [x] Field.Message for input example text that sits below field
- [x] Field.Valid for messages related to the validity of the input's value
- [x] Add Zod for more flexibility with validation rules
- [x] Custom onBlur in order to display/style Field.Valid
- [ ] Add logic to custom onBlur so that the message displays unique styling for error or warning or success. Only one message displays at a time prioritizing errors over warnings.
- [ ] State variable to track if input is empty (RHF isTouched?). This may need to be available globally so that we can provide a stepper component that allows user to set focus on empty inputs one at a time. Anticipate this would behave similar to cycling through git merge conflicts.
  - ![example_stepper_gitconflict](https://user-images.githubusercontent.com/48400779/233225509-4f2e2856-10a6-49b7-810b-7ea3a3417fe1.jpg)
- [ ] Add types 
- [ ] Refactor `Tip` component to be more generic and handle different message styles


#### Open Q's:
- To keep the logic simplified, should there be a separate WatchField component for conditional fields that will set their visibility based on another input's value?
  - For example, my first pass did this with `FieldTest` and `FieldTestConditional` 
- Should a top-level zod schema object be used or should the validation logic be passed at FIeld level
  - Top level seems advised as good practice but I'm concerned it makes it easy to make mistakes? 
  - Does this also create an issue with visibility conditions where the schema object needs nesting? See [example](https://www.reddit.com/r/reactjs/comments/10fxloj/how_do_you_use_reacthookform_and_zod_together/)
- It appears that validation libraries only support blocking form submission. It would be nice to have a middle ground where we warn the user but don't block form submission. Zod has [.optional](https://zod.dev/?id=optional) but it [doesn't seem to play well ](https://stackoverflow.com/questions/73715295/react-hook-form-with-zod-resolver-optional-field)with RHF and I couldn't get it to work.
- Is it better to use a single `Field.Message` that accepts a style prop to render messages in different styles (example, warning, error, success) or would it be better to create separate context components: `Field.Example`, `Field.Warning`, `Field.Error`, `Field.Success`

#### Examples:

##### `field-article.tsx`
1. Set up zod schema
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/pages/docs/field-article.tsx#L33-L51
2. Use `Field` with an atomic `Input` text component
  - Example 1:  https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/pages/docs/field-article.tsx#L122-L133
  - Example 2: https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/pages/docs/field-article.tsx#L135-L147

##### `field.tsx`
1. Setup `Field` and `FieldContext` at top level
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L9-L19
2. `Field.Control` uses RHF's controller to make the child component controlled. 
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L21-L45
3. Created `customOnBlur` in `Field.Control` to enable showing validation messages when focus leaves input
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L25-L28
  - Renamed defaultOnBlur and set logic so that the RHF's default onBlur is still available https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L37-L38
4. Use Radix UI `Slot` to pass `Controller` props to child component
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L39-L41
5. `Field.GroupLabel` uses Radix UI's `Label` to set an overall label for the input
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L47-L58
7. `FIeld.Tip`, `FIeld.Message`, `Field.Valid` are currently just the `Tip` component with a hardcoded prop
  - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L60-L77
  - A work in progress to try making the `Field.Valid` change message styling based on whether the message is an error or not
    - https://github.com/tayv/Project-Bubblegum/blob/914a8fb392b50107933b5dc80cd015032bb5d35f/designSystem/forms/Field.tsx#L79-L93
